### DISCUSSION TITLE
Updating Entity with Optional Relation

### DISCUSSION THREAD

**Discussion by edeesis** [micronaut-projects/micronaut-data#1978] [Q&A] [answered]
Hello, we are using R2dbc and have an entity similar to this:

```kotlin
@MappedEntity("entity")
data class Entity(
    val name: String,
    @field:Id val id: Long? = null
) {
  @Relation(Relation.Kind.MANY_TO_ONE) var relationshipOne: RelationOne? = null
  @Relation(Relation.Kind.MANY_TO_ONE) var relationshipTwo: RelationTwo? = null
}
```
The idea here is that one operation can initialize and save the Entity, another operation sets relationshipOne, and another operation sets relationshipTwo.

We do this by doing:
```kotlin
val entity = entityRepository.findBy(id)
entity.relationshipOne = RelationshipOne(...)
entityRepository.update(entity)
```

However, when we do this again in a second place:

```kotlin
val entity = entityRepository.findBy(id)
entity.relationshipTwo = RelationshipTwo(...)
entityRepository.update(entity)
```

The update statement generated sets relationship_one_id to null.

It seems like the issue is because we don't have @Join annotations on the Repository, so the Entity that gets loaded has relationshipOne and relationshipTwo set to null.

We worked around this for now by removing the Relation declarations, but that's not ideal obviously.

**Comment by dstepanov** [comment, marked as answer]
You can use something like `update(@Id Long id, RelationshipOne relationshipOne)` which will update only one field.

**Reply by edeesis** [reply]
Yeah. Thought of that, but then we'd have to warn people not to use the Repository#update method.

I wonder, is it possible to define both the foreign key (relationship_one_id) and the relation? Then the update wouldn't set it to null unless the FK is set to null

**Reply by dstepanov** [reply]
Not sure what you mean. The main thing to understand is that Micronaut Data creates SQL queries at the compilation time, so you cannot have runtime conditions to include/exclude some columns.

**Reply by edeesis** [reply]
I was meaning something like this:

```kotlin
@MappedEntity("entity")
data class Entity(
    val name: String,
    @field:Id val id: Long? = null
) {
  @Relation(Relation.Kind.MANY_TO_ONE) var relationshipOne: RelationOne? = null
  var relationshipOneId: Long? = null
  @Relation(Relation.Kind.MANY_TO_ONE) var relationshipTwo: RelationTwo? = null
  var relationshipTwoId: Long? = null
}
```

If relationshipOneId is set, and relationshipTwoId is null, and we call Repository#update, will the parameters of the update statement correctly be set, or will relationshipOne be set to null because it wasn't populated on load.

**Reply by dstepanov** [reply]
We generate the update query at the compilation time without knowing what values will be, so the query is always:
`UPDATE ... SET id, name, relationshipOneId, relationshipTwoId ...`
